bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java

```diff
@@ -77,27 +76,31 @@ public void processBindingConfiguration(String context, Item item, String bindin
 			throw new BindingConfigParseException("context is not permitted to be null for item "+item);
 		}
 
-		Set<Item> items = contextMap.get(context);
-		if (items==null) {
-			items = new HashSet<Item>();
-			contextMap.put(context, items);
+		synchronized (contextMap) {
+			Set<Item> items = contextMap.get(context);
+			if (items==null) {
+				items = new HashSet<Item>();
+				contextMap.put(context, items);
+			}
+
+			items.add(item);
 		}
-			
-		items.add(item);
 	}
 
 	/**
 	 * {@inheritDoc}
 	 */
 	public void removeConfigurations(String context) {
-		Set<Item> items = contextMap.get(context);
-		if(items!=null) {
-			for(Item item : items) {
-				// we remove all binding configurations for all items
-				bindingConfigs.remove(item.getName());
-				notifyListeners(item);
+		synchronized (contextMap) {
+			Set<Item> items = contextMap.get(context);
+			if(items!=null) {
+				for(Item item : items) {
+					// we remove all binding configurations for all items
+					bindingConfigs.remove(item.getName());
+					notifyListeners(item);
+				}
+				contextMap.remove(context);
 			}
```

Comment:
Thanks very much for having a look, @mrguessed.  I was hoping you would take an interest.  I will move the loop to call `notifyListeners` outside the lock as you suggest.  I didn't want to re-format the changes for fear it would hurt readability, but I see your point.  Any other thoughts/ideas/warnings most appreciated.